zip-0311.rst

```diff
@@ -2,7 +2,262 @@
+  improved if they do cache witness data for spent notes.)
+
+
+Specification
+=============
+
+Payment disclosure data structure
+---------------------------------
+
+A payment disclosure has the following fields:
+
+- `txid`: Transaction id for the given transaction `tx`.
+
+- `msg`: A message field, which could contain a challenge value from the party to whom
+  the payment disclosure is directed.
+
+- `sapling_outputs`: A vector of the Sapling Outputs that we are disclosing
+  [0..tx.shieldedOutputs.len()]
+
+  - `index`: An index into `tx.shieldedOutputs`.
+  - `ock`: The outgoing cipher key that allows this output to be recovered.
+
+- `sapling_spends`: A vector of the Sapling Spends for which we are proving spend authority
+  [0..tx.shieldedSpends.len()]
+
+  - `index`: An index into `tx.shieldedSpends`.
+  - `cv`
+  - `rk`
+  - `zkproof_spend`
+  - [Optional] A payment address proof:
+    - Any `(d, pk_d)` such that `pk_d = [ivk] GH(d)`
+    - `nullifier_addr`
+    - `zkproof_addr`
+  - `spendAuthSig`
+
+- `transparent_inputs`: A vector of the transparent inputs for which we are proving spend
+  authority [0..tx.vin.len()]
+
+  - `index`: An index into `tx.vin`.
+  - `sig`: A BIP 322 signature.
+
+    - TODO: `zcashd` currently only supports the legacy format defined in BIP 322. We may
+      want to backport full BIP 322 support before having transparent input support in
+      this ZIP, to ensure it does what we need.
+    - TODO: BIP 322 specifies consensus rule checks as part of the signature verification
+      process. We will likely need to migrate it over to an equivalent ZIP that specifies
+      these for Zcash (which has a different set of script validation consensus rules).
+
+Creating a payment disclosure
+-----------------------------
+
+The inputs to a payment disclosure are:
+
+- The transaction.
+- The SLIP-44 [#slip-0044]_ coin type.
+- The message :math:`msg` to be included (which may be empty).
+- A vector of `(output_index, ock)` tuples.
+- A vector of tuples of:
+
+  - A Sapling spend index.
+  - Its corresponding expanded spending key :math:`(\mathsf{ask}, \mathsf{nsk}, \mathsf{ovk})`.
+  - (Optional) An associated payment address :math:`(\mathsf{d}, \mathsf{pk_d})`.
+
+- A vector of tuples of:
+
+  - `index`: An index into `tx.vin`.
+  - The inputs to a BIP 322 signature (excluding `message_data`).
+
+The payment disclosure is created as follows:
+
+- For each Sapling spend index:
+
+  - Create a Sapling spend proof for the note that was spent in `tx.shieldedSpends[index]`,
+    using the same anchor, to obtain `(cv, rk, zkproof_spend)` as well as the random α
+    that was generated internally.
+
+  - [Optional] If an associated payment address was provided for this spend index, create
+    a ZIP 304 signature proof for that payment address, using α and `rk` from the
+    previous step. We obtain `(nullifier_addr, zkproof_addr)` from this step.
```

Comment:
Say that `nullifier_addr` will be the same nullifier that was revealed in the original spend.

Reply:
That is not the case. This nullifier is for the ZIP 304 fake note; we already use the nullifier from the original spend in `zkproof_spend`.